// factory/PageFactory.js
var esut = require('easy_util');
var digestUtil = esut.digestUtil;

var control = require('control');
var adminPageControl = control.adminPageCtrl;

var PageFactory = function(){};

PageFactory.prototype.handle = function(headNode, bodyNode, cb)
{
    console.log(bodyNode);
    var self = this;
    var cmd = headNode.cmd;
    self[cmd[0]](headNode, bodyNode, function(err, backBodyNode){
        backBodyNode.frameId = bodyNode.frameId;
        backBodyNode.parentFrameId = bodyNode.parentFrameId;
        cb(err, backBodyNode);
    });
}

PageFactory.prototype.admin = function(headNode, bodyNode, cb)
{
    adminPageControl.handle(headNode, bodyNode, cb);
}

PageFactory.prototype.team = function(headNode, bodyNode, cb)
{
    control.teamCtrl.handle(headNode, bodyNode, cb);
}

PageFactory.prototype.stage = function(headNode, bodyNode, cb)
{
    control.stageCtrl.handle(headNode, bodyNode, cb);
}

PageFactory.prototype.season = function(headNode, bodyNode, cb)
{
    control.seasonCtrl.handle(headNode, bodyNode, cb);
}

PageFactory.prototype.league = function(headNode, bodyNode, cb)
{
    control.leagueCtrl.handle(headNode, bodyNode, cb);
}

PageFactory.prototype.league = function(headNode, bodyNode, cb)
{
    control.leagueCtrl.handle(headNode, bodyNode, cb);
}

PageFactory.prototype.match = function(headNode, bodyNode, cb)
{
    control.matchCtrl.handle(headNode, bodyNode, cb);
}

var fac = new PageFactory();
module.exports = fac;
